ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/InitiatorBase.java

```diff
@@ -0,0 +1,302 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.common.ValidReadTxnList;
+import org.apache.hadoop.hive.common.ValidTxnList;
+import org.apache.hadoop.hive.common.ValidWriteIdList;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.*;
+import org.apache.hadoop.hive.metastore.metrics.AcidMetricService;
+import org.apache.hadoop.hive.metastore.metrics.Metrics;
+import org.apache.hadoop.hive.metastore.metrics.MetricsConstants;
+import org.apache.hadoop.hive.metastore.txn.CompactionInfo;
+import org.apache.hadoop.hive.metastore.txn.TxnCommonUtils;
+import org.apache.hadoop.hive.metastore.txn.TxnUtils;
+import org.apache.hadoop.hive.metastore.utils.JavaUtils;
+import org.apache.hadoop.hive.ql.ddl.table.storage.compact.AlterTableCompactDesc;
+import org.apache.hadoop.hive.ql.io.AcidDirectory;
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+import org.apache.hadoop.hive.ql.metadata.Partition;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.shims.HadoopShims;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hive.common.util.Ref;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.apache.hadoop.hive.ql.io.AcidUtils.compactionTypeStr2ThriftType;
+
+public class InitiatorBase extends MetaStoreCompactorThread{
+
+  static final private String COMPACTORTHRESHOLD_PREFIX = "compactorthreshold.";
+
+  public InitiatorBase() {
+  }
+
+  public InitiatorBase(Table table, List<Partition> partitions, AlterTableCompactDesc desc, HiveConf hiveConf)
+      throws Exception {
+    setConf(hiveConf);
+    AtomicBoolean flag = new AtomicBoolean();
+    init(flag);
+    initiateCompactionForTable(table,partitions,desc);
+  }
+
+  private void initiateCompactionForTable(Table table, List<Partition> partitions, AlterTableCompactDesc desc){
```

Comment:
You should move `AlterTableCompactOperation#compact` inside initator. 

1. non-partitioned and single partition compacts initiated in the Operator
2. multi partition compacts initiated via Initiator
3. The logic is differing: oderbyclause, compaction pool, etc are not used in the new case.

The logic to differentiate between the cases and the way new compactions requested should be common and compaction modifiers should not be disregarded for multi-partition compacts.

Reply:
New changes:

- if single partition or not partitioned table create a single request
- if multi partition request, create multiple requests according to the logic (thresholds, etc)

initiateCompactionForTable method accepts the AlterTableCompactDesc object and rest of the logic is inside that method leaving the operation class clean as we have moved all the initiation related code inside this new class.